test/gtest/test_multi.cc

```diff
@@ -0,0 +1,615 @@
+  int pages = std::get<1>(GetParam());
+
+  auto size = page_size() * pages;
+  data_ptr = mmap(size);
+  data_size = size;
+
+  pattern_fill(data_ptr, size, seed);
+
+  auto fork_action = [&](void* ptr) {
+    // Don't trigger page fault
+  };
+  StartAttach(peers, data_ptr, size, fork_action);
+
+  Share()->segid[0] = segid;
+  SignalAll();
+  SignalAll();
+}
+
+TEST_P(test_multi_xpmem_peers_pages, simple_one_to_many_no_parent_fault) {
+  int peers = std::get<0>(GetParam());
+  int pages = std::get<1>(GetParam());
+
+  auto size = page_size() * pages;
+  data_ptr = mmap(size);
+  data_size = size;
+
+  // Don't fill the memory area
+
+  auto fork_action = [&](void* ptr) {
+    // Read fault, but pattern won't be there
+    ASSERT_FALSE(test_pattern_check(ptr, size, seed));
+  };
+  StartAttach(peers, data_ptr, size, fork_action);
+
+  Share()->segid[0] = segid;
+  SignalAll();
+  SignalAll();
+}
+
+TEST_P(test_multi_xpmem_peers_pages, simple_one_to_many_sparse) {
+  int peers = std::get<0>(GetParam());
+  int pages = std::get<1>(GetParam());
+
+  int total_pages = 100 * 1000;
+  if (pages > total_pages) {
+    GTEST_SKIP();
+  }
+  auto size = page_size() * total_pages;
+  data_ptr = mmap(size);
+  data_size = size;
+
+  for (int i = 0; i < size; i += page_size()) {
+    auto input = reinterpret_cast<int*>(reinterpret_cast<char*>(data_ptr) + i);
+
+    *input = i + seed;
+  }
+
+  auto fork_action = [&](void* ptr) {
+    auto output = reinterpret_cast<int*>(ptr);
+    for (int i = 0; i < pages * page_size(); i += page_size()) {
+      auto output = reinterpret_cast<int*>(reinterpret_cast<char*>(ptr) + i);
+      EXPECT_EQ(i + seed, *output);
+    }
+  };
+  StartAttach(peers, data_ptr, size, fork_action);
+
+  Share()->segid[0] = segid;
+  SignalAll();
+  SignalAll();
+}
+
+TEST_P(test_multi_xpmem_peers_pages, simple_one_to_many) {
+  int peers = std::get<0>(GetParam());
+  int pages = std::get<1>(GetParam());
+
+  auto size = page_size() * pages;
+  data_ptr = mmap(size);
+  data_size = size;
+
+  pattern_fill(data_ptr, size, seed);
+
+  auto fork_action = [&](void* ptr) {
+    ASSERT_TRUE(test_pattern_check(ptr, size, seed));
+  };
+  StartAttach(peers, data_ptr, size, fork_action);
+
+  Share()->segid[0] = segid;
+  SignalAll();
+  SignalAll();
+}
+
+TEST_P(test_multi_xpmem_peers_pages, parallel_many_to_one) {
+  int peers = std::get<0>(GetParam());
+  int pages = std::get<1>(GetParam());
+
+  auto size = page_size() * pages;
+  data_ptr = mmap(size);
+  data_size = size;
+
+  auto fork_action = [&](void* ptr) {
+    auto data = static_cast<uint8_t*>(ptr);
+    for (int i = 0; i < size; i += page_size() / 2) {
+      seed = pattern_next(seed);
+      data[i] = seed;  // Uses lowest byte only
+    }
+  };
+  StartAttach(peers, data_ptr, size, fork_action);
+
+  WaitForAll();  // Wait for them to be ready
+  Share()->segid[0] = segid;
+  SignalAll();  // Start fork action
+  SignalAll();  // Waits for them
```

Comment:
can we define a func/method which calls `SignallAll` twice and use it everywhere?

Reply:
fixed